File: tests/runtime/records/host-wasmer.rs
use anyhow::Result;
use wasmer::WasmerEnv;

wit_bindgen_wasmer::export!("./tests/runtime/records/imports.wit");

use imports::*;

#[derive(WasmerEnv, Clone)]
pub struct MyImports;

impl Imports for MyImports {
    fn multiple_results(&mut self) -> (u8, u16) {
        (4, 5)
    }

    fn swap_tuple(&mut self, a: (u8, u32)) -> (u32, u8) {
        (a.1, a.0)
    }

    fn roundtrip_flags1(&mut self, a: F1) -> F1 {
        drop(a.to_string());
        drop(format!("{:?}", a));
        drop(a & F1::all());
        a
    }

    fn roundtrip_flags2(&mut self, a: F2) -> F2 {
        a
    }

    fn roundtrip_flags3(
        &mut self,
        a: Flag8,
        b: Flag16,
        c: Flag32,
        d: Flag64,
    ) -> (Flag8, Flag16, Flag32, Flag64) {
        (a, b, c, d)
    }

    fn roundtrip_record1(&mut self, a: R1) -> R1 {
        drop(format!("{:?}", a));
        a
    }

    fn tuple0(&mut self, _: ()) {}

    fn tuple1(&mut self, a: (u8,)) -> (u8,) {
        (a.0,)
    }
}

wit_bindgen_wasmer::import!("./tests/runtime/records/exports.wit");

fn run(wasm: &str) -> Result<()> {
    use exports::*;

    let exports = crate::instantiate(
        wasm,
        |store, import_object| imports::add_to_imports(store, import_object, MyImports),
        |store, module, import_object| exports::Exports::instantiate(store, module, import_object),
    )?;

    exports.test_imports()?;
    assert_eq!(exports.multiple_results()?, (100, 200));
    assert_eq!(exports.swap_tuple((1u8, 2u32))?, (2u32, 1u8));
    assert_eq!(exports.roundtrip_flags1(F1::A)?, F1::A);
    assert_eq!(exports.roundtrip_flags1(F1::empty())?, F1::empty());
    assert_eq!(exports.roundtrip_flags1(F1::B)?, F1::B);
    assert_eq!(exports.roundtrip_flags1(F1::A | F1::B)?, F1::A | F1::B);

    assert_eq!(exports.roundtrip_flags2(F2::C)?, F2::C);
    assert_eq!(exports.roundtrip_flags2(F2::empty())?, F2::empty());
    assert_eq!(exports.roundtrip_flags2(F2::D)?, F2::D);
    assert_eq!(exports.roundtrip_flags2(F2::C | F2::E)?, F2::C | F2::E);

    let r = exports.roundtrip_record1(R1 {
        a: 8,
        b: F1::empty(),
    })?;
    assert_eq!(r.a, 8);
    assert_eq!(r.b, F1::empty());

    let r = exports.roundtrip_record1(R1 {
        a: 0,
        b: F1::A | F1::B,
    })?;
    assert_eq!(r.a, 0);
    assert_eq!(r.b, F1::A | F1::B);

    assert_eq!(exports.tuple0(())?, ());
    assert_eq!(exports.tuple1((1,))?, (1,));
    Ok(())
}

File: tests/runtime/records/host-wasmer.py
from exports.bindings import Exports
from imports.bindings import add_imports_to_imports, Imports
from typing import Tuple, Any
import exports.bindings as e
import imports.bindings as i
import sys
import wasmer # type: ignore

class MyImports:
    def multiple_results(self) -> Tuple[int, int]:
        return (4, 5)

    def swap_tuple(self, a: Tuple[int, int]) -> Tuple[int, int]:
        return (a[1], a[0])

    def roundtrip_flags1(self, a: i.F1) -> i.F1:
        return a

    def roundtrip_flags2(self, a: i.F2) -> i.F2:
        return a

    def roundtrip_flags3(self, a: i.Flag8, b: i.Flag16, c: i.Flag32, d: i.Flag64) -> Tuple[i.Flag8, i.Flag16, i.Flag32, i.Flag64]:
        return (a, b, c, d)

    def roundtrip_record1(self, a: i.R1) -> i.R1:
        return a

    def tuple0(self, a: None) -> None:
        pass

    def tuple1(self, a: Tuple[int]) -> Tuple[int]:
        return (a[0],)

def run(wasm_file: str) -> None:
    store = wasmer.Store()
    module = wasmer.Module(store, open(wasm_file, 'rb').read())
    wasi_version = wasmer.wasi.get_version(module, strict=False)
    if wasi_version is None:
        import_object = {}
    else:
        wasi_env = wasmer.wasi.StateBuilder('test').finalize()
        import_object = wasi_env.generate_imports(store, wasi_version)

    wasm: Exports
    def get_export(name: str) -> Any:
        return wasm.instance.exports.__getattribute__(name)

    imports = MyImports()
    add_imports_to_imports(store, import_object, imports, get_export)
    wasm = Exports(store, import_object, module)

    wasm.test_imports()
    assert(wasm.multiple_results() == (100, 200))
    assert(wasm.swap_tuple((1, 2)) == (2, 1))
    assert(wasm.roundtrip_flags1(e.F1.A) == e.F1.A)
    assert(wasm.roundtrip_flags1(e.F1(0)) == e.F1(0))
    assert(wasm.roundtrip_flags1(e.F1.A | e.F1.B) == (e.F1.A | e.F1.B))

    assert(wasm.roundtrip_flags2(e.F2.C) == e.F2.C)
    assert(wasm.roundtrip_flags2(e.F2(0)) == e.F2(0))
    assert(wasm.roundtrip_flags2(e.F2.D) == e.F2.D)
    assert(wasm.roundtrip_flags2(e.F2.C | e.F2.E) == (e.F2.C | e.F2.E))

    r = wasm.roundtrip_record1(e.R1(8, e.F1(0)))
    assert(r.a == 8)
    assert(r.b == e.F1(0))

    r = wasm.roundtrip_record1(e.R1(a=0, b=e.F1.A | e.F1.B))
    assert(r.a == 0)
    assert(r.b == (e.F1.A | e.F1.B))

    wasm.tuple0(None)
    assert(wasm.tuple1((1,)) == (1,))

if __name__ == '__main__':
    run(sys.argv[1])
